Libplanet.Net/BlockCandidateTable.cs

```diff
@@ -10,88 +10,121 @@
+            _table = new ConcurrentDictionary<BlockHeader, List<Block<T>>>();
         }
 
         public long Count
         {
-            get => _blocks.Count;
+            get => _table.Count;
         }
 
-        public bool Any() => !_blocks.IsEmpty;
-
         /// <summary>
-        /// Adds a <see cref="Block{T}"/>s to the table.
+        /// <para>
+        /// Adds given <paramref name="blocks"/> to the table.
+        /// </para>
+        /// <para>
+        /// The internal table is only updated if a given pair satisfy all of the following:
+        /// <list type="bullet">
+        ///     <item><description>
+        ///         The internal dictionary does not already contain <paramref name="blockHeader"/>
+        ///         as its key.
+        ///     </description></item>
+        ///     <item><description>
+        ///         Given <paramref name="blocks"/> is non-empty.
+        ///     </description></item>
+        ///     <item><description>
+        ///         Given <paramref name="blocks"/> are consecutive in the sense that indices
+        ///         are unique consecutive and every <see cref="Block{T}.PreviousHash"/> match
+        ///         the <see cref="Block{T}.Hash"/> of the previous <see cref="Block{T}"/>.
+        ///     </description></item>
+        /// </list>
+        /// </para>
         /// </summary>
         /// <param name="blockHeader">This is the header of the <see cref="BlockChain{T}"/>
         /// tip at the time of downloading the blocks.</param>
-        /// <param name="blocks">List of downloaded <see cref="Block{T}"/>.</param>
+        /// <param name="blocks">The list of downloaded <see cref="Block{T}"/>s.</param>
         public void Add(BlockHeader blockHeader, IEnumerable<Block<T>> blocks)
         {
-            if (_blocks.ContainsKey(blockHeader))
+            if (_table.ContainsKey(blockHeader))
             {
+                Log.Debug(
+                    "Given blocks will not be added as the table already contains " +
+                    "blockheader #{Index} {BlockHash} as a key",
+                    blockHeader.Index,
+                    blockHeader.Hash);
                 return;
             }
 
-            try
+            List<Block<T>> sorted = blocks.OrderBy(block => block.Index).ToList();
+            if (!sorted.Any())
             {
-                var sortedBlocks =
-                    new SortedList<long, Block<T>>(blocks.ToDictionary(i => i.Index));
-                if (sortedBlocks.Count > 0)
-                {
-                    _blocks.TryAdd(blockHeader, sortedBlocks);
-                }
+                Log.Debug(
+                    "Given blocks will not be added to the table as an empty set of blocks " +
+                    "is not allowed",
+                    sorted.Count);
+                return;
             }
-            catch (ArgumentException e)
+            else if (!sorted
+                .Zip(sorted.Skip(1), (prev, next) =>
```

Comment:
Have you any reason for skipping the first item?

Reply:
Otherwise it would map `{ 1, 2, 3, 4 }` to `{ (1, 1), (2, 2), (3, 3), (4, 4) }`? 🙃